lib/semantics/semantics.cc

```diff
@@ -20,51 +20,52 @@
 #include "rewrite-parse-tree.h"
 #include "scope.h"
 #include "symbol.h"
+#include <ostream>
 
 namespace Fortran::semantics {
 
 static void DoDumpSymbols(std::ostream &, const Scope &, int indent = 0);
 static void PutIndent(std::ostream &, int indent);
 
-Semantics &Semantics::set_searchDirectories(
-    const std::vector<std::string> &directories) {
-  for (auto directory : directories) {
-    directories_.push_back(directory);
-  }
-  return *this;
-}
-
-Semantics &Semantics::set_moduleDirectory(const std::string &directory) {
-  moduleDirectory_ = directory;
-  directories_.insert(directories_.begin(), directory);
-  return *this;
+bool SemanticsContext::AnyFatalError() const {
+  return !messages_.empty() &&
+      (warningsAreErrors_ || messages_.AnyFatalError());
 }
 
-bool Semantics::Perform(parser::Program &program) {
-  ValidateLabels(messages_, program);
+bool Semantics::Perform() {
+  ValidateLabels(context_.messages(), program_);
   if (AnyFatalError()) {
     return false;
   }
-  parser::CanonicalizeDo(program);
-  ResolveNames(messages_, globalScope_, program, directories_, defaultKinds_);
+  parser::CanonicalizeDo(program_);
+  ResolveNames(context_, program_);
   if (AnyFatalError()) {
     return false;
   }
-  RewriteParseTree(messages_, globalScope_, program);
+  RewriteParseTree(context_, program_);
   if (AnyFatalError()) {
     return false;
   }
-  ModFileWriter writer;
-  writer.set_directory(moduleDirectory_);
-  if (!writer.WriteAll(globalScope_)) {
-    messages_.Annex(writer.errors());
+  ModFileWriter writer{context_};
+  writer.WriteAll();
+  if (AnyFatalError()) {
     return false;
   }
-  return true;
+  if (context_.debugExpressions()) {
+    parser::CharBlock whole{cooked_.data()};
+    parser::ContextualMessages contextualMessages{whole, &context_.messages()};
+    evaluate::FoldingContext foldingContext{contextualMessages};
+    AnalyzeExpressions(program_, foldingContext, context_);
```

Comment:
Should the FoldingContext be part of SemanticsContext. So far, nothing in SemanticsContext gets reset between source files, so FoldingContext would change that. If we include it, references to CookedSource and Program could also go in there.

Reply:
The folding context wraps some contextualized messages and some floating-point flags.  The flags won't change.  I believe that the context of the contextualized messages is pretty meaningless at the top level; new higher-resolution contexts are created on demand during expression semantics so that error messages have precise locations.  Does that help?